traffic_ops/traffic_ops_golang/parameter/parameters.go

```diff
@@ -0,0 +1,354 @@
+package parameter
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/apache/incubator-trafficcontrol/lib/go-log"
+	"github.com/apache/incubator-trafficcontrol/lib/go-tc"
+	"github.com/apache/incubator-trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"github.com/apache/incubator-trafficcontrol/traffic_ops/traffic_ops_golang/auth"
+	"github.com/apache/incubator-trafficcontrol/traffic_ops/traffic_ops_golang/dbhelpers"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/lib/pq"
+)
+
+//we need a type alias to define functions on
+type TOParameter tc.ParameterNullable
+
+//the refType is passed into the handlers where a copy of its type is used to decode the json.
+var refType = TOParameter(tc.ParameterNullable{})
+
+func GetRefType() *TOParameter {
+	return &refType
+}
+
+//Implementation of the Identifier, Validator interface functions
+func (parameter *TOParameter) GetID() (int, bool) {
+	if parameter.ID == nil {
```

Comment:
these shouldn't be pointer receivers -- only methods that modify the content should be..:
```
func (parameter TOParameter)...
```

Reply:
>these shouldn't be pointer receivers -- only methods that modify the content should be

I don't agree. Non-pointers copy the entire object on every call. There's no reason to create an expensive copy of this object, just to get the ID.

Reasons for value-functions are 
* when the object is small, so the safety of being unable to modify the real object outweighs the performance cost
* the function does modify the object for its own use, but it's necessary to leave the called object unmodified
* The object is small enough, and the function called frequently enough, and performance matters, that a copy is faster than the cache-locality-breaking pointer lookup.
* It's desirable to avoid fulfilling a pointer interface

None of those apply here.

That said, the object isn't _that_ big, I don't feel super-strongly that it needs to be a pointer, if you feel that strongly in the other direction.

But I disagree strongly with the statement "only methods that modify the content should be [pointer receivers]". Value member functions in Go create a copy of the entire object on every call, and that needs to be considered.

Reply:
yeah -- that was an over-generalization.   I agree that if the object is large that's a concern, but in these cases it's not.

Reply:
and it's honestly not worth haggling over.   I've reviewed it and this is the only thing I brought up.   merging..